Add a quirk to store third-party cookies in partitioned storage

szewai opened the pull request:
#### 24260b1c915eded6dcd704ae8dd8dc05723c2d75
<pre>
Add a quirk to store third-party cookies in partitioned storage
<a href="https://bugs.webkit.org/show_bug.cgi?id=270494">https://bugs.webkit.org/show_bug.cgi?id=270494</a>
<a href="https://rdar.apple.com/113830141">rdar://113830141</a>

Reviewed by Chris Dumez.

cagreatamerica.com does not display correctly because an iframe from queue-it.net checks to cookie storage access from
script, and third-party cookies are fully blocked when ITP is enabled. To fix that, allow queue-it.net to read and write
cookies in partitioned storage. Note the storage is partitioned by top-level domain, so quirked domain cannot perform
cross-site tracking. Also the storage currently resides in the memory of web process, so the cookies are not applied to
network tasks.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldUseEphemeralPartitionedStorageForDOMCookies const):
* Source/WebCore/page/Quirks.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebCookieJarCocoa.mm: Added.
(WebKit::policyProperties):
(WebKit::WebCookieJar::cookiesInPartitionedCookieStorage const):
(WebKit::WebCookieJar::setCookiesInPartitionedCookieStorage):
(WebKit::WebCookieJar::ensurePartitionedCookieStorage):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::cookies const):
(WebKit::WebCookieJar::setCookies):
(WebKit::WebCookieJar::cookiesInPartitionedCookieStorage const):
(WebKit::WebCookieJar::setCookiesInPartitionedCookieStorage):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:

Canonical link: <a href="https://commits.webkit.org/275763@main">https://commits.webkit.org/275763@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ddf580b2672eeadc67a608f8181b580dd61988f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42737 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21758 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45138 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45349 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38861 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45043 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25423 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19123 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35373 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43310 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18764 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36786 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16331 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/16394 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37846 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/793 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38907 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38177 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46857 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17555 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14473 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42098 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/19174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40727 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9541 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19353 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18819 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1838,4 +1838,19 @@ bool Quirks::shouldIgnorePlaysInlineRequirementQuirk() const
 #endif
 }
 
+bool Quirks::canUsePartitionedCookieStorage(const URL& url) const
```
The naming seems odd? Surely everyone can use a partitioned cookie storage?

szewai replied:
We currently blocks all third-party cookies, so technically no one can use partitioned cookies storage -- there is only non-partitioned cookie storage for first-party use.

cdumez replied:
I see. I hadn't thought of it this way. It makes more sense now butI still think we could come up with a clearer name.
Maybe we should indicate that this is an in-memory storage for e.g. ? Or that it only impact DOM cookies (i.e. document.cookie)

brentfulgham replied:
I think a better name for the Quirk is `shouldUsePartitionedCookieStorage(...)`

szewai replied:
Sure, I will rename it to "shouldUseEphemeralPartitionedStorageForDOMCookies"

cdumez commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebCookieJarCocoa.mm:
```diff
@@ -0,0 +1,113 @@
+
+    __block RetainPtr<NSArray> cookies;
+    [m_partitionedCookieStorage.get() _getCookiesForURL:cookieURL mainDocumentURL:firstPartyURL partition:partition policyProperties:policyProperties(sameSiteInfo, cookieURL).get() completionHandler:^(NSArray *result) {
+        cookies = retainPtr(result);
```
`retainPtr()` is likely not needed.

cdumez commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebCookieJarCocoa.mm:
```diff
@@ -0,0 +1,113 @@
+        cookies = retainPtr(result);
+    }];
+
+    if (!cookies || ![cookies.get() count])
```
`_getCookiesForURL` is async, no? It looks async but then it seems we expect a result synchronously so I am a bit confused.

szewai replied:
Surprisingly it is sync... see comment in `static RetainPtr<NSArray> cookiesForURL(NSHTTPCookieStorage *storage, NSURL *url, NSURL *mainDocumentURL, const std::optional<SameSiteInfo>& sameSiteInfo, NSString *partition = nullptr)`.

cdumez replied:
Ok, good to know. Just looked a bit weird.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.h:
```diff
@@ -76,9 +80,18 @@ class WebCookieJar final : public WebCore::CookieJar {
     HashMap<String, WeakHashSet<WebCore::CookieChangeListener>> m_changeListeners;
+
+#if PLATFORM(COCOA)
+    RetainPtr<NSHTTPCookieStorage> m_partitionedCookieStorage;
```
I think this needs a better name and maybe a comment.

cdumez replied:
Have you thought about the lifetime of these cookies? Seems this storage never gets cleared (or removed from). It also seems it is page wide and pages can get navigated from one domain to another? Do we care about the fact that cookies for a given domain stay in the process basically forever, even after you're done visiting this domain?

szewai replied:
I am not too concerned about the lifetime because the cookies are small, non-persistent and there is only one quirked domain now, and we swap process on navigation. Also the page could navigate back to the domain so we might want to keep them.

brentfulgham replied:
I believe these cookies exist for the lifetime of the WebContent process, not the Network Process, correct? If so, I agree the lifetime should not be a concern.

szewai replied:
Yes, this is only in the memory of web process and not used for network requests.

cdumez commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebCookieJarCocoa.mm:
```diff
@@ -0,0 +1,113 @@
+        if (![[cookie name] length] || [cookie isHTTPOnly])
+            continue;
+
+        cookiesBuilder.append(cookiesBuilder.isEmpty() ? "" : "; ", String([cookie name]), "=", String([cookie value]));
```
We should probably use `makeString(cookiesBuilder.isEmpty() ? "" : "; ", [cookie name], "=", [cookie value])` to avoid intermediate String allocations.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -161,11 +161,13 @@ void WebCookieJar::setCookies(WebCore::Document& document, const URL& url, const
     ApplyTrackingPrevention applyTrackingPreventionInNetworkProcess = ApplyTrackingPrevention::No;
-    if (shouldBlockCookies(webFrame.get(), document.firstPartyForCookies(), url, applyTrackingPreventionInNetworkProcess))
+    if (shouldBlockCookies(webFrame.get(), document.firstPartyForCookies(), url, applyTrackingPreventionInNetworkProcess)) {
+        setCookiesInPartitionedCookieStorage(document, url, sameSiteInfo, cookieString);
```
Maybe we can add the `IfNecessary` suffix here to clarify that it doesn't always happen. As a matter of fact, it rarely happens.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -391,4 +393,17 @@ void WebCookieJar::removeChangeListener(const String& host, const WebCore::Cooki
+{
+}
+
+#endif
```
`#endif // !PLATFORM(COCOA)`

cdumez reviewed (approved):
LGTM with nits.

szewai committed:
Add a quirk to store third-party cookies in partitioned storage
https://bugs.webkit.org/show_bug.cgi?id=270494
rdar://113830141

Reviewed by Chris Dumez.

cagreatamerica.com does not display correctly because an iframe from queue-it.net checks to cookie storage access from
script, and third-party cookies are fully blocked when ITP is enabled. To fix that, allow queue-it.net to read and write
cookies in partitioned storage. Note the storage is partitioned by top-level domain, so quirked domain cannot perform
cross-site tracking. Also the storage currently resides in the memory of web process, so the cookies are not applied to
network tasks.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldUseEphemeralPartitionedStorageForDOMCookies const):
* Source/WebCore/page/Quirks.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebCookieJarCocoa.mm: Added.
(WebKit::policyProperties):
(WebKit::WebCookieJar::cookiesInPartitionedCookieStorage const):
(WebKit::WebCookieJar::setCookiesInPartitionedCookieStorage):
(WebKit::WebCookieJar::ensurePartitionedCookieStorage):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::cookies const):
(WebKit::WebCookieJar::setCookies):
(WebKit::WebCookieJar::cookiesInPartitionedCookieStorage const):
(WebKit::WebCookieJar::setCookiesInPartitionedCookieStorage):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:

Canonical link: https://commits.webkit.org/275763@main
